### ISSUE TITLE
'PG::ConnectionBad: could not connect to server: No such file or directory' during run rake db:migrate 

### ISSUE THREAD

**Issue by taraszka** [flynn/flynn#4191]
Hey,

I am pretty sure that rake should take available ENV["DATABASE_URL"] instead of looking for local Unix domain socket during db:migrate action. 

Please have a look below:

flynn -c eu-west-1 -a testing-admin-1 run rake db:migrate --trace
** Invoke db:migrate (first_time)
** Invoke environment (first_time)
** Execute environment
** [Honeybadger] Starting Honeybadger version 2.1.3 level=1 pid=12
DataProtection: Application has found XYZ users
DataProtection: Data is marked as redacted
I, [2017-07-27T11:24:33.759452 #12]  INFO -- : Cannot access /app/log/sqreen.log for writing. Defaulting to stdout
** Invoke db:load_config (first_time)
** Execute db:load_config
** Execute db:migrate
rake aborted!
PG::ConnectionBad: could not connect to server: No such file or directory
	Is the server running locally and accepting
	connections on Unix domain socket "/var/run/postgresql/.s.PGSQL.5432"?
/app/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.8/lib/active_record/connection_adapters/postgresql_adapter.rb:651:in `initialize'
/app/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.8/lib/active_record/connection_adapters/postgresql_adapter.rb:651:in `new'
/app/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.8/lib/active_record/connection_adapters/postgresql_adapter.rb:651:in `connect'
/app/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.8/lib/active_record/connection_adapters/postgresql_adapter.rb:242:in `initialize'
/app/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.8/lib/active_record/connection_adapters/postgresql_adapter.rb:44:in `new'
/app/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.8/lib/active_record/connection_adapters/postgresql_adapter.rb:44:in `postgresql_connection'
/app/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.8/lib/active_record/connection_adapters/abstract/connection_pool.rb:438:in `new_connection'
/app/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.8/lib/active_record/connection_adapters/abstract/connection_pool.rb:448:in `checkout_new_connection'
/app/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.8/lib/active_record/connection_adapters/abstract/connection_pool.rb:422:in `acquire_connection'
/app/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.8/lib/active_record/connection_adapters/abstract/connection_pool.rb:349:in `block in checkout'
/app/vendor/ruby-2.2.6/lib/ruby/2.2.0/monitor.rb:211:in `mon_synchronize'
/app/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.8/lib/active_record/connection_adapters/abstract/connection_pool.rb:348:in `checkout'
/app/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.8/lib/active_record/connection_adapters/abstract/connection_pool.rb:263:in `block in connection'
/app/vendor/ruby-2.2.6/lib/ruby/2.2.0/monitor.rb:211:in `mon_synchronize'
/app/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.8/lib/active_record/connection_adapters/abstract/connection_pool.rb:262:in `connection'
/app/vendor/bundle/ruby/2.2.0/gems/ar-octopus-0.8.5/lib/octopus/proxy.rb:208:in `safe_connection'
/app/vendor/bundle/ruby/2.2.0/gems/ar-octopus-0.8.5/lib/octopus/proxy.rb:212:in `select_connection'
/app/vendor/bundle/ruby/2.2.0/gems/ar-octopus-0.8.5/lib/octopus/proxy.rb:273:in `method_missing'
/app/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.8/lib/active_record/migration.rb:916:in `initialize'
/app/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.8/lib/active_record/migration.rb:823:in `new'
/app/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.8/lib/active_record/migration.rb:823:in `up'
/app/vendor/bundle/ruby/2.2.0/gems/ar-octopus-0.8.5/lib/octopus/migration.rb:110:in `up_with_octopus'
/app/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.8/lib/active_record/migration.rb:801:in `migrate'
/app/vendor/bundle/ruby/2.2.0/gems/ar-octopus-0.8.5/lib/octopus/migration.rb:104:in `block in migrate_with_octopus'
/app/vendor/bundle/ruby/2.2.0/gems/ar-octopus-0.8.5/lib/octopus/proxy.rb:226:in `block (2 levels) in run_queries_on_shard'
/app/vendor/bundle/ruby/2.2.0/gems/ar-octopus-0.8.5/lib/octopus/proxy.rb:476:in `using_shard'
/app/vendor/bundle/ruby/2.2.0/gems/ar-octopus-0.8.5/lib/octopus/proxy.rb:225:in `block in run_queries_on_shard'
/app/vendor/bundle/ruby/2.2.0/gems/ar-octopus-0.8.5/lib/octopus/proxy.rb:462:in `keeping_connection_proxy'
/app/vendor/bundle/ruby/2.2.0/gems/ar-octopus-0.8.5/lib/octopus/proxy.rb:224:in `run_queries_on_shard'
/app/vendor/bundle/ruby/2.2.0/gems/ar-octopus-0.8.5/lib/octopus/proxy.rb:233:in `block in send_queries_to_multiple_shards'
/app/vendor/bundle/ruby/2.2.0/gems/ar-octopus-0.8.5/lib/octopus/proxy.rb:232:in `each'
/app/vendor/bundle/ruby/2.2.0/gems/ar-octopus-0.8.5/lib/octopus/proxy.rb:232:in `send_queries_to_multiple_shards'
/app/vendor/bundle/ruby/2.2.0/gems/ar-octopus-0.8.5/lib/octopus/migration.rb:103:in `migrate_with_octopus'
/app/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.8/lib/active_record/tasks/database_tasks.rb:137:in `migrate'
/app/vendor/bundle/ruby/2.2.0/gems/activerecord-4.2.8/lib/active_record/railties/databases.rake:44:in `block (2 levels) in <top (required)>'
/app/vendor/bundle/ruby/2.2.0/gems/rake-10.5.0/lib/rake/task.rb:240:in `call'
/app/vendor/bundle/ruby/2.2.0/gems/rake-10.5.0/lib/rake/task.rb:240:in `block in execute'
/app/vendor/bundle/ruby/2.2.0/gems/rake-10.5.0/lib/rake/task.rb:235:in `each'
/app/vendor/bundle/ruby/2.2.0/gems/rake-10.5.0/lib/rake/task.rb:235:in `execute'
/app/vendor/bundle/ruby/2.2.0/gems/rake-10.5.0/lib/rake/task.rb:179:in `block in invoke_with_call_chain'
/app/vendor/ruby-2.2.6/lib/ruby/2.2.0/monitor.rb:211:in `mon_synchronize'
/app/vendor/bundle/ruby/2.2.0/gems/rake-10.5.0/lib/rake/task.rb:172:in `invoke_with_call_chain'
/app/vendor/bundle/ruby/2.2.0/gems/rake-10.5.0/lib/rake/task.rb:165:in `invoke'
/app/vendor/bundle/ruby/2.2.0/gems/rake-10.5.0/lib/rake/application.rb:150:in `invoke_task'
/app/vendor/bundle/ruby/2.2.0/gems/rake-10.5.0/lib/rake/application.rb:106:in `block (2 levels) in top_level'
/app/vendor/bundle/ruby/2.2.0/gems/rake-10.5.0/lib/rake/application.rb:106:in `each'
/app/vendor/bundle/ruby/2.2.0/gems/rake-10.5.0/lib/rake/application.rb:106:in `block in top_level'
/app/vendor/bundle/ruby/2.2.0/gems/rake-10.5.0/lib/rake/application.rb:115:in `run_with_threads'
/app/vendor/bundle/ruby/2.2.0/gems/rake-10.5.0/lib/rake/application.rb:100:in `top_level'
/app/vendor/bundle/ruby/2.2.0/gems/rake-10.5.0/lib/rake/application.rb:78:in `block in run'
/app/vendor/bundle/ruby/2.2.0/gems/rake-10.5.0/lib/rake/application.rb:176:in `standard_exception_handling'
/app/vendor/bundle/ruby/2.2.0/gems/rake-10.5.0/lib/rake/application.rb:75:in `run'
/app/bin/rake:4:in `<main>'
Tasks: TOP => db:migrate

Then I ran rake console to see if I have maybe wrongly set the DATABASE_URL but it was just fine:
irb(main):003:0> ENV["DATABASE_URL"]
=> "postgres://USER:PASSWORD@testing-admin-1.AN_AWS.eu-west-1.rds.amazonaws.com:5432/my_app"

I ran db:migrate inside rake console and got same issue:

irb(main):001:0> %x[rake db:migrate]
rake aborted!
PG::ConnectionBad: could not connect to server: No such file or directory
	Is the server running locally and accepting
	connections on Unix domain socket "/var/run/postgresql/.s.PGSQL.5432"?

It's look like a bug.

**Comment by lmars** [comment]
@taraszka the scheme in your `DATABASE_URL` should be `postgresql`, so:

```
postgresql://USER:PASSWORD@testing-admin-1.AN_AWS.eu-west-1.rds.amazonaws.com:5432/my_app
```

**Comment by taraszka** [comment]
hmm. Ok. I will check this. By default Flynn ships with postgres:

$ flynn -a testing-www-2 env | grep DATABASE_URL
DATABASE_URL=postgres://abf0ea05992abf606c9ba8545cdbd747:3227826be057fbd873e72bffa1fe6674@leader.postgres.discoverd:5432/dae6ce4a4892a8add485870d357bccf0

if it should be with postgresql then I think it should be replaced in Flynn I guess.

**Comment by taraszka** [comment]
Ok. I've tried with both: postgresql:// and postgres:// and still got the same. It want to talk to local Unix domain socket rather than remote database for rake db:migration task. 
@lmars - it sounds like a bug.

**Comment by titanous** [comment]
The correct scheme is `postgres://`, however your app is not using the `DATABASE_URL` variable. This is a problem with your app not with Flynn, because as you point out, the env var is set correctly.

**Comment by taraszka** [comment]
@titanous I don't think so. The same works fine with PG provided by Flynn.

**Comment by taraszka** [comment]
@titanous and at Heroku ;)

**Comment by titanous** [comment]
I'm unclear on what you're asking? As you can see in the Rails console, the variable is being set by Flynn. So the variable is making it into your app. If your app is not using the variable, then it is likely that there is a configuration or code issue that is causing it to not work correctly. If you can provide a simple reproduction case that demonstrates a bug in Flynn, we'd love to investigate and fix it.

**Comment by taraszka** [comment]
Okay. So I you could see above my testing-admin-1 app was not able to use 'rake db:migrate' task or rake console to call db:migrate when the database endpoint was set to RDS or any external system. 

Same setup but with database deployed at Flynn works just fine:

$ flynn -c eu-west-1 -a testing-admin-2 run rake db:migrate --trace
** Invoke db:migrate (first_time)
** Invoke environment (first_time)
** Execute environment
** [Honeybadger] Starting Honeybadger version 2.1.3 level=1 pid=15
DataProtection: Application has found XYZ users
DataProtection: Data is marked as redacted
I, [2017-07-27T16:21:48.188843 #15]  INFO -- : Cannot access /app/log/sqreen.log for writing. Defaulting to stdout
** Invoke db:load_config (first_time)
** Execute db:load_config
** Execute db:migrate
** Invoke db:_dump (first_time)
** Execute db:_dump
** Invoke db:schema:dump (first_time)
** Invoke environment 
** Invoke db:load_config 
** Execute db:schema:dump

and below please see console and DATABASE_URL env:

$ flynn -c eu-west-1 -a testing-admin-2 run rails console
** [Honeybadger] Starting Honeybadger version 2.1.3 level=1 pid=15
DataProtection: Application has found XYZ users
DataProtection: Data is marked as redacted
I, [2017-07-27T16:25:07.947890 #15]  INFO -- : Cannot access /app/log/sqreen.log for writing. Defaulting to stdout
Loading staging environment (Rails 4.2.8)
irb(main):001:0> ENV["DATABASE_URL"]
=> "postgres://abf0ea05992abf606c9ba8545cdbd747:3227826be057fbd873e72bffa1fe6674@leader.postgres.discoverd:5432/dae6ce4a4892a8add485870d357bccf0"
irb(main):002:0> %x[rake db:migrate]
=> "** [Honeybadger] Starting Honeybadger version 2.1.3 level=1 pid=26\nDataProtection: Application has found XYZ users\nDataProtection: Data is marked as redacted\nI, [2017-07-27T16:25:29.201422 #26]  INFO -- : Cannot access /app/log/sqreen.log for writing. Defaulting to stdout\n"

To summarize it, when I've set an external database using DATABASE_URL rake tried to use local unix socket:

PG::ConnectionBad: could not connect to server: No such file or directory
Is the server running locally and accepting
connections on Unix domain socket "/var/run/postgresql/.s.PGSQL.5432"?

For me It's look like a bug and that is why I've reported it here.

**Comment by titanous** [comment]
Is it possible that your app is using the `PGHOST`, `PGUSER`, `PGPASSWORD`, and `PGDATABASE` env vars that Flynn sets in the first example? As I said before, all Flynn is doing is setting the environment variables, what your app chooses to do with them is up to it.

**Comment by taraszka** [comment]
Hmmm. that is valid hint. Let me check.

**Comment by taraszka** [comment]
Ok. Thanks @titanous. I've unset PGHOST PGPASSWORD PGUSER PGDATABASE and left only DATABASE_URL point to internal database and had the same.

I will close this incident as invalid. Thanks for your great support and clarification.
